// Copyright (c) farm-ng, inc.
//
// Licensed under the Amiga Development Kit License (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://github.com/farm-ng/amiga-dev-kit/blob/main/LICENSE
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Copyright (c) farm-ng, inc. All rights reserved.

syntax = "proto3";


import "farm_ng/core/lie.proto";
import "farm_ng/core/linalg.proto";
import "farm_ng/core/pose.proto";

package farm_ng.control.proto;


message ControllerState {
  // The pose of the robot in the world frame
  farm_ng.core.proto.Pose world_from_robot = 1;
  // The pose of the current goal in the robot frame
  farm_ng.core.proto.VecXF64 robot_from_goal = 2;
  // The commanded velocity in the robot's egocentric frame
  farm_ng.core.proto.Isometry3F64Tangent commands = 3;

  // Rotation about the z-axis contained in the world_from_robot pose, in radians
  double heading = 4;

  // Which track are we following
  string track_name = 5;
  // The closest point on the track in the world frame
  farm_ng.core.proto.Pose world_from_track = 6;
  // The pose of the current goal in the world frame
  farm_ng.core.proto.Pose world_from_goal = 7;

  // The distance of the robot from the closest point along the track
  // This is the difference between world_from_robot and world_from_track
  farm_ng.core.proto.VecXF64 robot_from_track = 8;

  // The index of the current goal along the track
  int32 current_index = 9;
  // The number of goals along the track
  int32 track_size = 10;
  // Whether the robot is following a track or not
  bool following_track = 11;
  // If the amiga dashboard is in auto mode or not.
  bool auto_mode = 12;
}

message ControllerTrack {
  // The name of the track
  string name = 1;
  // The poses of the track
  repeated farm_ng.core.proto.Pose poses = 2;
}
